File: L300-usecase/action-center/backend/lakebase.py
"""Lakebase (managed Postgres) access — the SHARED write-back module.

Reused verbatim by the sibling Supervisor and Finance apps for their own
write-back tables (forecast_overrides, scm_interventions, commercial_actions,
agent_sessions, ...). This is the *write* governance plane: a short-lived DB
credential (the app/service write identity) generated via the SDK, psycopg3,
sslmode=require, search_path set to the `akzo` schema.

The credential is short-lived (~1h), so we cache it and transparently refresh on
expiry. `query()` / `execute()` open a fresh autocommit connection per call —
simple and correct for an app's low write volume; swap in a pool later if needed.
"""
from __future__ import annotations

import os
import threading
import time
from contextlib import contextmanager
from typing import Any

import psycopg
from psycopg import sql as psql
from psycopg.rows import dict_row

from databricks_client import client, current_user

INSTANCE_NAME = os.environ.get("LAKEBASE_INSTANCE", "")
DB_NAME = os.environ.get("LAKEBASE_DBNAME", "databricks_postgres")
PG_SCHEMA = os.environ.get("LAKEBASE_SCHEMA", "akzo")

_lock = threading.Lock()
_host: str | None = None
_token: str | None = None
_token_exp: float = 0.0
# Refresh a little before the ~1h expiry to avoid mid-request failures.
_TOKEN_TTL_SECONDS = 45 * 60


def _read_write_host() -> str:
    global _host
    if _host is None:
        with _lock:
            if _host is None:
                inst = client().database.get_database_instance(name=INSTANCE_NAME)
                _host = inst.read_write_dns
    return _host


def _db_token() -> str:
    """Short-lived Lakebase credential — the app/service write identity. Cached + refreshed."""
    global _token, _token_exp
    now = time.time()
    if _token is None or now >= _token_exp:
        with _lock:
            # Re-check inside the lock (another thread may have just refreshed).
            if _token is None or time.time() >= _token_exp:
                cred = client().database.generate_database_credential(instance_names=[INSTANCE_NAME])
                _token = cred.token
                _token_exp = now + _TOKEN_TTL_SECONDS
    return _token


@contextmanager
def connection():
    """Open an autocommit psycopg connection with search_path set to the akzo schema.

    Rows come back as dicts. Transparently retries once with a fresh token if the
    cached credential has expired.
    """
    def _connect() -> psycopg.Connection:
        return psycopg.connect(
            host=_read_write_host(), port=5432, dbname=DB_NAME,
            user=current_user(), password=_db_token(), sslmode="require",
            autocommit=True, row_factory=dict_row,
        )

    try:
        conn = _connect()
    except psycopg.OperationalError:
        # Likely an expired token — force a refresh and retry once.
        global _token
        _token = None
        conn = _connect()

    try:
        with conn.cursor() as cur:
            cur.execute(
                psql.SQL("SET search_path TO {}").format(psql.Identifier(PG_SCHEMA))
            )
        yield conn
    finally:
        conn.close()


def query(sql: str, params: tuple | None = None) -> list[dict[str, Any]]:
    """Run a read query, return list-of-dicts."""
    with connection() as conn, conn.cursor() as cur:
        cur.execute(sql, params)
        return cur.fetchall()


def execute(sql: str, params: tuple | None = None, returning: bool = False):
    """Run a write. If returning=True, fetch and return the first row (a dict)."""
    with connection() as conn, conn.cursor() as cur:
        cur.execute(sql, params)
        if returning:
            return cur.fetchone()
        return None

File: hackathon-starter-kit/hackathon-hub/server/lakebase.ts
/**
 * Lakebase (managed Postgres) access for the Hackathon hub — the Node port of
 * apps/_shared/lakebase.py. AppKit's `lakebase` plugin targets the newer Postgres
 * *projects* API; the workshop uses a classic Database Instance (set via
 * LAKEBASE_INSTANCE), so we connect directly with `pg` + a short-lived OAuth credential.
 *
 * Identity: whoever the Databricks SDK config authenticates as — the developer's
 * email locally, the app service principal when deployed. The credential is the
 * pg password; it expires ~1h, so we cache + refresh transparently. search_path
 * is pinned to the configured schema where every workshop write-back table lives.
 *
 * All connection details come from env vars (see .env.example); nothing is
 * hardcoded so the hub is portable across workspaces and Vocareum labs.
 */
import pg from 'pg';
import type { Pool } from 'pg';
import { randomUUID } from 'node:crypto';
import { Config } from '@databricks/sdk-experimental';

const INSTANCE = process.env.LAKEBASE_INSTANCE;
if (!INSTANCE) throw new Error('LAKEBASE_INSTANCE env var is required');
const HOST = process.env.PGHOST;
if (!HOST) throw new Error('PGHOST env var is required (Lakebase Postgres endpoint host)');
const DB_NAME = process.env.PGDATABASE ?? process.env.LAKEBASE_DBNAME ?? 'databricks_postgres';
const SCHEMA = process.env.LAKEBASE_SCHEMA ?? 'akzo';
// Refresh a little before the ~1h expiry to avoid mid-request failures.
const TOKEN_TTL_MS = 45 * 60 * 1000;

let _cfg: Config | null = null;
async function config(): Promise<Config> {
  if (!_cfg) {
    const cfg = new Config({ profile: process.env.DATABRICKS_CONFIG_PROFILE });
    await cfg.ensureResolved();
    _cfg = cfg;
  }
  return _cfg;
}

async function authHeader(): Promise<string> {
  const h = new Headers();
  await (await config()).authenticate(h);
  const auth = h.get('Authorization');
  if (!auth) throw new Error('Databricks authentication produced no Authorization header');
  return auth;
}

let _user: string | null = null;
/** The identity the SDK authenticates as (email locally, SP when deployed). */
export async function currentUser(): Promise<string> {
  if (_user) return _user;
  const cfg = await config();
  const res = await fetch(`${cfg.host}/api/2.0/preview/scim/v2/Me`, {
    headers: { Authorization: await authHeader() },
  });
  if (!res.ok) throw new Error(`SCIM Me failed: ${res.status} ${await res.text()}`);
  const me = (await res.json()) as { userName?: string };
  if (!me.userName) throw new Error('SCIM Me returned no userName');
  _user = me.userName;
  return _user;
}

let _token: string | null = null;
let _tokenExp = 0;
async function dbToken(): Promise<string> {
  if (_token && Date.now() < _tokenExp) return _token;
  const cfg = await config();
  const res = await fetch(`${cfg.host}/api/2.0/database/credentials`, {
    method: 'POST',
    headers: { Authorization: await authHeader(), 'Content-Type': 'application/json' },
    body: JSON.stringify({ instance_names: [INSTANCE], request_id: randomUUID() }),
  });
  if (!res.ok) throw new Error(`generate-database-credential failed: ${res.status} ${await res.text()}`);
  const cred = (await res.json()) as { token: string };
  _token = cred.token;
  _tokenExp = Date.now() + TOKEN_TTL_MS;
  return _token;
}

// node-postgres pool. The async password provider is invoked per new connection,
// so an expired cached credential is refreshed before the next connect; search_path
// is set on each new connection so every query lands in the akzo schema.
let _pool: Pool | null = null;
export async function getPool(): Promise<Pool> {
  if (_pool) return _pool;
  const user = await currentUser();
  const pool = new pg.Pool({
    host: HOST,
    port: 5432,
    database: DB_NAME,
    user,
    password: dbToken,
    ssl: { rejectUnauthorized: false },
    max: 5,
    idleTimeoutMillis: 30_000,
    connectionTimeoutMillis: 10_000,
  });
  pool.on('connect', (client) => {
    void client.query(`SET search_path TO ${SCHEMA}`);
  });
  _pool = pool;
  return _pool;
}

/** Run a query and return the rows. Params use $1, $2, ... placeholders. */
export async function query<T = Record<string, unknown>>(
  text: string,
  params: unknown[] = []
): Promise<T[]> {
  const pool = await getPool();
  const res = await pool.query(text, params);
  return res.rows as T[];
}

/** Run a write and return the first RETURNING row (or null). */
export async function execute<T = Record<string, unknown>>(
  text: string,
  params: unknown[] = []
): Promise<T | null> {
  const rows = await query<T>(text, params);
  return rows[0] ?? null;
}

export const lakebaseSchema = SCHEMA;
